src/replication.c

```diff
@@ -2386,6 +2481,353 @@ char *sendCommandArgv(connection *conn, int argc, char **argv, size_t *argv_lens
     return NULL;
 }
 
+sds getReplicaPortString(void) {
+    long long replica_port;
+    if (server.slave_announce_port) {
+        replica_port = server.slave_announce_port;
+    } else if (server.tls_replication && server.tls_port) {
+        replica_port = server.tls_port;
+    } else {
+        replica_port = server.port;
+    }
+    return sdsfromlonglong(replica_port);
+}
+
+void freeReplDataBuf(void) {
+    if (server.pending_repl_data.blocks) {
+        listRelease(server.pending_repl_data.blocks);
+    }
+    server.pending_repl_data.blocks = NULL;
+    server.pending_repl_data.len = 0;
+}
+
+void abortRdbConnectionSync(int should_retry) {
+    if (server.repl_transfer_s) {
+        connClose(server.repl_transfer_s);
+        server.repl_transfer_s = NULL;
+    }
+    if (server.repl_full_sync_s) {
+        connClose(server.repl_full_sync_s);
+        server.repl_full_sync_s = NULL;
+    }
+    if (server.psync_master) {
+        freeClient(server.psync_master);
+        server.psync_master = NULL;
+    }
+    if (server.repl_transfer_fd != -1) {
+        close(server.repl_transfer_fd);
+        bg_unlink(server.repl_transfer_tmpfile);
+        zfree(server.repl_transfer_tmpfile);
+        server.repl_transfer_tmpfile = NULL;
+        server.repl_transfer_fd = -1;
+    }
+    server.repl_transfer_tmpfile = NULL;
+    server.repl_transfer_fd = -1;
+    freeReplDataBuf();
+
+    if (!should_retry) {
+        /* Make sure next time we will try different approach */
+        server.primary_can_sync_using_rdb_channel = 0;
+    }
+    return;
+}
+
+int reInitReplicaMainConnection(connection *conn) {
+    serverAssert(conn == server.repl_transfer_s);
+    /* Init the main connection for psync */
+    conn->state = CONN_STATE_CONNECTED;
+    server.repl_state = REPL_SEC_CONN_SEND_PSYNC;
+    serverAssert(connSetReadHandler(server.psync_master->conn, syncWithMaster) != C_ERR);
+    syncWithMaster(conn);
+    return C_OK;
+}
+
+int prepareRdbConnectionForRdbLoad(connection *conn) {
+    serverAssert(isRdbConnectionSync(conn));
+    /* Set connection handler for RDB */
+    serverAssert(connSetReadHandler(conn, readSyncBulkPayload) != C_ERR);
+    server.repl_transfer_size = -1;
+    server.repl_transfer_read = 0;
+    server.repl_transfer_last_fsync_off = 0;
+    server.repl_transfer_lastio = server.unixtime;
+    return C_OK;
+}
+
+int processEndOffsetResponse(char* response) {
+    long long reploffset;
+    char master_replid[CONFIG_RUN_ID_SIZE+1];
+    int dbid;
+    char endoff_format[30];
+    snprintf(endoff_format, sizeof(endoff_format), "$ENDOFF:%%lld %%%ds %%d", CONFIG_RUN_ID_SIZE);
```

Comment:
Though we probably won't change CONFIG_RUN_ID_SIZE, asserting that it is 40 seems more confusing than snprintf. My preference is to keep this line general, and to add documentation.

Reply:
you can also use a STRINGIFY macro and let the compiler concatenate adjacent strings..
but again, this constant is part of the protocol, it will never change, it can't change.

it's like remaining flexible in case TCP will some day no longer be IP protocol number 6.

Reply:
Ack, I'll replace it with the actual value next revision